File: idea-plugin/src/main/java/org/ifinalframework/jetbrains/plugins/aio/api/yapi/DefaultYapiService.java
package org.ifinalframework.jetbrains.plugins.aio.api.yapi;


import org.ifinalframework.jetbrains.plugins.aio.api.yapi.model.*;
import org.springframework.stereotype.Service;

import javax.annotation.Resource;
import java.util.List;

/**
 * DefaultYapiService
 *
 * @author iimik
 * @since 0.0.1
 **/
@Service
class DefaultYapiService implements YapiService {
    @Resource
    private YapiProperties properties;
    @Resource
    private YapiClient yapiClient;

    @Override
    public Long getProjectId() {
        final Result<Project> projectInfo = yapiClient.getProjectInfo(properties.getToken());
        if (projectInfo.isSuccess()) {
            return projectInfo.getData().getId();
        }
        return null;
    }

    @Override
    public Long getCatMenuId(String category) {
        // 1. find project id
        final Long projectId = getProjectId();
        if (projectId == null) {
            return null;
        }
        final Result<List<CatMenu>> result = yapiClient.getCatMenus(projectId, properties.getToken());
        if (!result.isSuccess()) {
            return null;
        }
        // 2. find cat id
        return result.getData().stream()
                .filter(it -> it.getName().equals(category))
                .findFirst().map(CatMenu::getId).orElse(null);

    }

    @Override
    public Api getApi(String category, String method, String path) {
        final Long catMenuId = getCatMenuId(category);
        if (catMenuId == null) {
            return null;
        }
        // 3. find api
        final Result<Page<Api>> result = yapiClient.getApiListInCat(properties.getToken(), catMenuId, 1, Integer.MAX_VALUE);
        if (!result.isSuccess()) {
            return null;
        }
        return result.getData().getList().stream()
                .filter(it -> it.getMethod().equalsIgnoreCase(method) && it.getPath().equalsIgnoreCase(path))
                .findFirst()
                .orElse(null);
    }
}

File: idea-plugin/src/main/kotlin/org/ifinalframework/jetbrains/plugins/aio/api/yapi/DefaultYapiService.kt
package org.ifinalframework.jetbrains.plugins.aio.api.yapi;

import com.itangcent.common.utils.firstOrNull
import jakarta.annotation.Resource
import org.ifinalframework.jetbrains.plugins.aio.api.yapi.model.Api
import org.ifinalframework.jetbrains.plugins.aio.api.yapi.model.CatMenu
import org.ifinalframework.jetbrains.plugins.aio.api.yapi.model.Project
import org.springframework.stereotype.Service


/**
 * DefaultYapiService
 *
 * @author iimik
 * @since 1.6.0
 **/
@Service
class DefaultYapiService : YapiService {
    @Resource
    private lateinit var yapiProperties: YapiProperties

    @Resource
    private lateinit var yapiClient: YapiClient

    override fun getProject(): Project? {
        return yapiClient.getProject(yapiProperties.token!!).data
    }

    override fun getCatMenu(category: String): CatMenu? {
        val project = getProject() ?: return null
        val result = yapiClient.getCatMenus(project.id, yapiProperties.token!!)
        return if (result.isSuccess()) {
            result.data!!.stream()
                .filter { it!!.name == category }
                .firstOrNull()
        } else null
    }

    override fun getApi(category: String, method: String, path: String): Api? {
        val catMenu = getCatMenu(category) ?: return null
        val result = yapiClient.getApiListInCat(yapiProperties.token!!, catMenu.id)
        return if (result.isSuccess()) {
            result.data.list.stream()
                .filter { it.method == method && it.path == path }
                .firstOrNull()
        } else null
    }
}